### DISCUSSION TITLE
Temporary failure in name resolution issue

### DISCUSSION THREAD

**Discussion by nkciam** [engageub/InternetIncome#50] [General]
Hi guys,

I faced the issue of "Temporary failure in name resolution" when I ran more than 600 containers. (including tun2sock and repocket and traff container). The issue appears when container number 601 is created and the other container creates an issue from this time. No device added from this point. Any solution for this one?

Thank you!

![image](https://github.com/engageub/InternetIncome/assets/70848927/a982f2c9-01ae-4fa0-9895-f73cd1df2400)

Thank Engageub a lot for creating this script!

**Comment by engageub** [comment]
Hi,
Your network may be throttling the DNS request since all the requests for DNS goes through the direct connection or it may have slowed down due to multiple requests which resulted in timeout. 
Here are a few reasons why DNS errors might occur when running a large number of containers:

DNS Server Capacity: The default DNS server provided by Docker might not be able to handle a high volume of DNS requests from a large number of containers. This can result in timeouts or failures to resolve hostnames.

DNS Caching: Docker employs DNS caching to improve performance, but a large number of containers can overwhelm the cache, leading to inconsistent or outdated DNS resolution.

Network Congestion: Running a high volume of containers can lead to network congestion, causing delays or timeouts in DNS requests.

Resource Constraints: Insufficient resources such as CPU, memory, or network bandwidth can impact DNS resolution, especially under heavy loads.

To confirm further, you can run the following command in terminal and see if your DNS server is responding when you run more than 600 containers. 
```
nslookup google.com 8.8.8.8
````
If your socks5 server supports UDP, this can be modified in the code to redirect DNS traffic through your proxies instead of direct connection. 
In addition to this, docker has a default limitation of 1023 containers on a single network. This can be resolved by creating different network and starting the other containers on a different network bridge. https://stackoverflow.com/questions/50394234/maximum-number-of-deployable-docker-containers

For the docker network limitation, code has to be modified to support multiple network bridges.

Thank you

**Reply by nkciam** [reply]
Here is the nslookup result:

![image](https://github.com/engageub/InternetIncome/assets/70848927/2e52c1a0-dfea-4361-a46c-e24702c79110)

Hi @engageub , thank you for your explanation. All clear to me. I knew docker has a default limitation of 1023 containers but when I run with your script with 1000 proxy for 2 app repocket and traffmonetizer. It created 4000 containers (2000 containers for tun2sock, 1000 for traff, 1000 for repocket) from your script. Over limit 1023 containers can created :D

**Reply by engageub** [reply]
Thank you for the response. Was the command run when the resolution errors occurred or after stopping it?
Regarding the containers, with 1000 proxy we are creating 1000 tun2socks containers on same network. The rest of the containers go through the tun2socks network and not direct network bridge. So they are not considered under the count for number of containers on network bridge. 
With 1000 proxies did you create 1000 tun2socks containers or 2000 tun2socks containers?

**Reply by nkciam** [reply]
@engageub  Yes. The command run when the resolution errors occurred.

With 1000 proxies the script created 2000 tun2socks containers if we run 2 apps like traff or repocket. (Total 4000 containers i checked by command)

Maybe with 1 network interface we should run only 500-600 container. (Included tun2sock and apps containers) to avoid issue above. But it good if we can adding more network interface. (For example with Azure VM we can add more NIC network interface).

Thank you!

**Reply by engageub** [reply]
@nkciam , Yes multiple networks can be added. The initial version published has network creation. Later it was removed since it was not necessary and only required if more than 1023 proxies are used.
However, you can create a new network using "docker network create" command and add that network in the code to tun2socks command and run the other 1000 proxies.
This feature will soon to be added to test branch to create custom network if necessary, for users who are willing to use more than 1023 proxies on a single host.

Thank you

**Reply by hieunguyen8794** [reply]
![image](https://github.com/engageub/InternetIncome/assets/25577311/9f851c2a-9280-4f5c-b503-d11bdd55f33d)
Has this feature been added yet? I can only run up to 1023 containers for traffmonetizer and 1023 for tunsock.

**Reply by engageub** [reply]
Hello,
This feature has not been added yet. 
However, I would suggest you to do the following for now.

1. Create docker network using the following command with network name as network1. You can keep incrementing the numbers for each network like network2,network3 etc for every 1023 containers.

```
sudo docker network create network1
```

2. After creating the network you have to use this network name in the script. Open internetIncome.sh file and go to line [number 265](https://github.com/engageub/InternetIncome/blob/test/internetIncome.sh#L265) and replace the line as mentioned below. We have added an extra parameter --network=network1 in the code below.

```
      docker_parameters=($LOGS_PARAM $MAX_MEMORY_PARAM $MEMORY_RESERVATION_PARAM $CPU_PARAM --network=network1 -e LOGLEVEL=$TUN_LOG_PARAM -e PROXY=$proxy -e EXTRA_COMMANDS="$EXTRA_COMMANDS" -v '/dev/net/tun:/dev/net/tun' --cap-add=NET_ADMIN $combined_ports xjasonlyu/tun2socks:v2.5.2)

```

Create separate folders for every 1023 proxies and you can use as many proxies as long as the networks are created.

**Reply by engageub** [reply]
Let me know if that's working fine after making changes.

**Comment by engageub** [comment]
Hi,
This feature is added to test branch now.
Please set USE_CUSTOM_NETWORK=true in the properties.conf at the end of the file.
Please confirm if this is working fine with more than 1023 proxies in the script. 
Also, try to use DNS_OVER_HTTPS or SOCKS5_DNS to avoid DNS throttling on the host.

Thank you

**Comment by hieunguyen8794** [comment]
I tested it, but because the number of containers is too large, the performance is really not good. but the way you say it ok, it works fine

**Reply by engageub** [reply]
You will also have to monitor the CPU while running with many proxies.
Could you please let me know if you see multiple networks in networks.txt . If so, it split the proxies with different networks.

Thank you

**Reply by engageub** [reply]
Hi,
Please also have a look at the recently updated [FAQ](https://github.com/engageub/InternetIncome?tab=readme-ov-file#-why-are-some-of-the-containers-for-the-same-application-offline) related to DNS queries.

Thank you
